### PR TITLE
Doc: apigroups, alpha, beta, experimental/v1alpha1

### PR DESCRIPTION
**Pull request by erictune**
My attempt to summarize ongoing discussions about versions and stability for a user audience.

### CODE REVIEW

**Comment by erictune** [comment]
@bgrant0607 @mikedanese @davidopp @quinton-hoole@aronchick @brendandburns @roberthbailey @smarterclayton are just some people who may be interested.

**Comment by bgrant0607** [comment]
Thanks. Will take a look.

File: docs/api.md [hunk 1]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+   - The version names contain `beta` (e.g. `v2beta3`).
+   - The object schema is not expected to change significantly in the next API version, but occasionally it
+     might.
+   - We will give advice and/or tools for migrating your objects from one beta version to the next
+     beta version and to the stable version.
+   - We expect overlapping support for "previous" and "current" version for several software releases,
+     to allow for migrating from one version to the next.
+   - Beta level objects/features are expected to graduate to Stable, and will not be dropped.
+   - We may not be able to make significant changes in the API once the API reaches the beta stage.
+     possible during the Alpha level.
```
Comment (roberthbailey) [inline docs/api.md]: sentence fragment.

File: docs/api.md [hunk 2]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+     even if you do not create alpha versioned objects.
+   - Please do try alpha versions in test clusters and give feedback, as significant changes are
+     possible during the Alpha level.
+ - Beta level:
+   - The version names contain `beta` (e.g. `v2beta3`).
+   - The object schema is not expected to change significantly in the next API version, but occasionally it
+     might.
+   - We will give advice and/or tools for migrating your objects from one beta version to the next
+     beta version and to the stable version.
+   - We expect overlapping support for "previous" and "current" version for several software releases,
```
Comment (roberthbailey) [inline docs/api.md]: I'm not sure we expect it for several releases (since releases are 3 months apart).

File: docs/api.md [hunk 3]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
 
 What constitutes a compatible change and how to change the API are detailed by the [API change document](devel/api_changes.md).
 
+
+## API groups
+
+To make it easier to extend the Kubernetes API, we are in the process of implementing *API groups*.
```
Comment (bgrant0607) [inline docs/api.md]: Link to proposals/api-group.md

File: docs/api.md [hunk 4]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+   `apiVersion: v1`.
+1. the "experimental" group, which is at REST path `/apis/experimental/$VERSION`, and which uses
+  `apiVersion: experimental/$VERSION`, where currently `apiVersion: experimental/v1alpha1`.
+  The name `experimental` implies that API groups themselves are currently under development.  The actual
+  stability of the version of objects within that group is determined by its [API
+  version](#api-versioning)
+
+In the future we expect that there will be more API groups, all at REST path `/apis/$API_GROUP` and
+using `apiVersion: $API_GROUP/$VERSION`.  We expect that there will be a way for third parties to
+create their own API groups, and to avoid naming collisions.
```
Comment (bgrant0607) [inline docs/api.md]: Link to design/extending-api.md

File: docs/api.md [hunk 5]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+
+Not all kubernetes clusters will support all API groups.  However, we intend that if an API server
```
Comment (bgrant0607) [inline docs/api.md]: Kubernetes (capitalized)

File: docs/api.md [hunk 6]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+supports a group, it supports all the Kinds in the group, and all of them at the same version(s).
+
 ## API versioning
 
-To make it easier to eliminate fields or restructure resource representations, Kubernetes supports multiple API versions, each at a different API path prefix, such as `/api/v1beta3`. These are simply different interfaces to read and/or modify the same underlying resources. In general, all API resources are accessible via all API versions, though there may be some cases in the future where that is not true.
+To make it easier to eliminate fields or restructure resource representations, Kubernetes supports multiple API versions, each at a different API path prefix, such as `/api/v1beta3`.
```
Comment (bgrant0607) [inline docs/api.md]: Versions are discussed in the discussion about groups. This explanation probably needs to be moved up.

You didn't like the sentence about accessing the same underlying resources?

File: docs/api.md [hunk 7]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
 
 We chose to version at the API level rather than at the resource or field level to ensure that the API presents a clear, consistent view of system resources and behavior, and to enable controlling access to end-of-lifed and/or experimental APIs.
 
-The [API and release versioning proposal](design/versioning.md) describes the current thinking on the API version evolution process.
+Note that API versioning and Software versioning are only indirectly related.  The [API and release
+versioning proposal](design/versioning.md) describes the current thinking on the API version
+evolution process.
```
Comment (bgrant0607) [inline docs/api.md]: That doc is about software versioning and the relationship between API versioning and software versioning. You're describing the API evolution process here.

File: docs/api.md [hunk 8]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+
+Different API versions imply different levels of support and stability:
```
Comment (bgrant0607) [inline docs/api.md]: "stability and support" sounds better to me.

File: docs/api.md [hunk 9]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+   - The object schema may change significantly in the next API version.
+   - We may not give any tools or instructions for migrating your objects from one version to the next
+     &emdash; you may have to delete, rewrite, and re-create your objects when you upgrade your cluster to the next
+     version.
+   - Support for object/feature may be dropped entirely in next software release.
+   - Do not count on overlapping support for "previous" and "current" version; supported version
+     may change abruptly on a new software release.
+   - Flag changes may be needed to enable alpha versions.
+   - Hosted clusters (e.g. GKE) may not support alpha versions.
+   - Code that implements alpha objects may be less well tested.  Enabling alpha support may active bugs
```
Comment (bgrant0607) [inline docs/api.md]: s/active/trigger/

File: docs/api.md [hunk 10]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+     even if you do not create alpha versioned objects.
+   - Please do try alpha versions in test clusters and give feedback, as significant changes are
```
Comment (bgrant0607) [inline docs/api.md]: Please put this above all of the caveats. Also mention that alpha features are under active development.

File: docs/api.md [hunk 11]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+   - Beta level objects/features are expected to graduate to Stable, and will not be dropped.
+   - We may not be able to make significant changes in the API once the API reaches the beta stage.
+     possible during the Alpha level.
+- Stable level:
+    - The version name is `vX` where `X` is an integer.
+    - Stable versions will be supported for a long time.
+
+## Enabling `experimental/v1alpha1`
+
+Enable `experimental/v1alpha1` objects by adding the following flags to your API server:
```
Comment (bgrant0607) [inline docs/api.md]: Can link to the doc in #15004 once that merges.

File: docs/api.md [hunk 12]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
-The [API and release versioning proposal](design/versioning.md) describes the current thinking on the API version evolution process.
+Note that API versioning and Software versioning are only indirectly related.  The [API and release
+versioning proposal](design/versioning.md) describes the current thinking on the API version
+evolution process.
+
+Different API versions imply different levels of support and stability:
+ - Alpha level:
+   - The version names contain `alpha` (e.g. `v1alpha1`).
+   - The object schema may change significantly in the next API version.
+   - We may not give any tools or instructions for migrating your objects from one version to the next
```
Comment (bgrant0607) [inline docs/api.md]: May not be fully implemented. For instance, some API operations, CLI commands, or UI support may not be implemented.

File: docs/api.md (hunk 2 above)
Comment (bgrant0607) [inline docs/api.md]: Yes, at least one software release.

**Comment by bgrant0607** [comment]
Captures the main points. Thanks.

File: docs/api.md (hunk 1 above)
Comment (erictune) [inline docs/api.md]: fixed

File: docs/api.md (hunk 2 above)
Comment (erictune) [inline docs/api.md]: Clarified.

File: docs/api.md (hunk 3 above)
Comment (erictune) [inline docs/api.md]: fixed

File: docs/api.md (hunk 4 above)
Comment (erictune) [inline docs/api.md]: fixed.

File: docs/api.md (hunk 5 above)
Comment (erictune) [inline docs/api.md]: Fixed

File: docs/api.md (hunk 6 above)
Comment (erictune) [inline docs/api.md]: Moved up.  I didn't like the sentence about *all API resources are accessible via all API* since that is no longer true.  I didn't mean to delete *These are simply different interfaces...underlying resources* and have reinstated that sentence.

File: docs/api.md (hunk 7 above)
Comment (erictune) [inline docs/api.md]: PTAL.

File: docs/api.md (hunk 8 above)
Comment (erictune) [inline docs/api.md]: ok

File: docs/api.md (hunk 9 above)
Comment (erictune) [inline docs/api.md]: ok

File: docs/api.md (hunk 10 above)
Comment (erictune) [inline docs/api.md]: PTAL

File: docs/api.md (hunk 11 above)
Comment (erictune) [inline docs/api.md]: That seems redundant with this one.

File: docs/api.md (hunk 12 above)
Comment (erictune) [inline docs/api.md]: fixed

**Comment by erictune** [comment]
@lavalamp this is trying to do the same thing as your #15004.

File: docs/api.md [hunk 13]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+To make it easier to extend the Kubernetes API, we are in the process of implementing *API groups*.
+The API group is specified in a REST path and in the `apiVersion` field of a serialized object.
+
+Currently there are two API groups in use:
+
+1. the "core" group, which is at REST path `/api/v1` and is not specified as part of the `apiVersion` field, e.g.
+   `apiVersion: v1`.
+1. the "experimental" group, which is at REST path `/apis/experimental/$VERSION`, and which uses
+  `apiVersion: experimental/$VERSION`, where currently `apiVersion: experimental/v1alpha1`.
+  The name `experimental` implies that API groups themselves are currently under development.  The actual
```
Comment (quinton-hoole) [inline docs/api.md]: These two sentences are still very confusing to me.  Needs further wordsmithing, IMO.  Will come back with some suggestions once I've grokked the rest of the doc.

File: docs/api.md [hunk 14]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+  stability of the version of objects within that group is determined by its [API
+  version](#api-versioning)
+
+In the future we expect that there will be more API groups, all at REST path `/apis/$API_GROUP` and
+using `apiVersion: $API_GROUP/$VERSION`.  We expect that there will be a way for third parties to
+create their own API groups, and to avoid naming collisions.
+
+Not all kubernetes clusters will support all API groups.  However, we intend that if an API server
+supports a group, it supports all the Kinds in the group, and all of them at the same version(s).
```
Comment (quinton-hoole) [inline docs/api.md]: I assume that this implies a dependency graph between Kubernetes applications and the API groups and versions that they depend upon? The same presumably applies between API groups themselves.  So for example  ApplicationX/VersionY depends upon APIGroupA/VersionB, which depends on APIGroupC/VersionD... etc.  So to move such an application between Kubernetes clusters requires that the entire dependency graph be compatible between the two clusters.  If I'm correct here, we need to be explicit about this, I think.

File: docs/api.md [hunk 15]
```diff
@@ -53,13 +53,79 @@ In our experience, any system that is successful needs to grow and change as new
...
+   - The version names contain `alpha` (e.g. `v1alpha1`).
+   - The object schema may change significantly in the next API version.
+   - We may not give any tools or instructions for migrating your objects from one version to the next
+     &emdash; you may have to delete, rewrite, and re-create your objects when you upgrade your cluster to the next
+     version.
+   - Support for object/feature may be dropped entirely in next software release.
+   - Do not count on overlapping support for "previous" and "current" version; supported version
+     may change abruptly on a new software release.
+   - Flag changes may be needed to enable alpha versions.
+   - Hosted clusters (e.g. GKE) may not support alpha versions.
```
Comment (quinton-hoole) [inline docs/api.md]: Perhaps a more specific word than 'support' here.  There's a difference between "your API calls will fail" vs "customer support won't answer your questions".

File: docs/api.md (hunk 14 above)
Comment (erictune) [inline docs/api.md]: I think that you only build "real" applications on top of "stable" versions, which last a long time, and you endeavor to keep all your clusters on the same set of (APIGroup, StableVersion) pairs.  You only enable alpha/beta groups on test clusters, or with the knowledge that you are going to have some porting work to move things.

File: docs/api.md (hunk 9 above)
Comment (quinton-hoole) [inline docs/api.md]: I think 'activate' is what you meant, which is slightly better then 'trigger', IMO.

File: docs/api.md (hunk 14 above)
Comment (erictune) [inline docs/api.md]: Would like to leave that problem for another PR.

**Comment by erictune** [comment]
Updated based on all comments and based on todays meeting.

**Comment by bgrant0607** [comment]
experimental/v1alpha1 is going to be renamed to extensions/v1beta1

File: docs/api.md (hunk 6 above)
Comment (bgrant0607) [inline docs/api.md]: I previously made a number of comments about linking to other docs. Did you decide not to do that?

The new api paths are at a different location: /apis/extensions/v1beta1. It would be useful to show that, also.

**Comment by smarterclayton** [comment]
Good summaries - I didn't have any serious issues with the statements made.

File: docs/api.md [hunk 16]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+features are under active development, you may wish to try alpha features in a test cluster, rather
+than in a production cluster.
+
+Different API versions imply different levels of stability and support:
+ - Alpha level:
+   - The version names contain `alpha` (e.g. `v1alpha1`).
+   - *Purpose*: sharing new features with developers and eager expert users, prior to:
+     - *complete implementation*: for instance, some API operations, CLI
+     commands, or UI support may not be implemented. 
+     - *thorough API review*: the object schema may change in a later software release, and it may be necessary
```
Comment (bgrant0607) [inline docs/api.md]: schema and/or semantics

While APIs are being developed, we break them a number of times, across all API versions. Existing objects may no longer function the same / correctly.

File: docs/api.md (hunk 6 above)
Comment (bgrant0607) [inline docs/api.md]: Nevermind about the links -- found them, just not where I expected them.

**Comment by bgrant0607** [comment]
Much better now, thanks.

File: docs/admin/daemons.md [hunk 17]
```diff
@@ -209,9 +209,12 @@ all or certain hosts, and when it needs to start before other pods.
 
 ## Caveats
 
-DaemonSet is part of the experimental API group, so it is not subject to the same compatibility
-guarantees as objects in the main API.  It may not be enabled.  Enable by setting
-`--runtime-config=experimental/v1alpha1` on the apiserver.
+DaemonSet objects are in the [`experimental` API Group](../api.md#api-groups).  This API group may or
+may not be enabled on your cluster.
```
Comment (davidopp) [inline docs/admin/daemons.md]: Should you say something about how to enable/disable it, or know if it is enabled/disabled?

File: docs/api.md [hunk 18]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+       to delete/modify/re-create objects in order to upgrade a cluster to a subsequent software release.
+       The object modification may not be easily automated, and the process may be poorly documented.
+     - *commitment to complete the feature*: the feature may be dropped entirely in a later software release.
+     - *allow rapid changes in versions*: API versions may increment faster than the minor release cadence.
+   - Alpha features are present in released binaries, but disabled by default. Enabling an Alpha feature
```
Comment (davidopp) [inline docs/api.md]: s/Enabling/When enabling/

File: docs/api.md [hunk 19]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+ - Beta level:
+   - The version names contain `beta` (e.g. `v2beta3`).
+   - *Purpose*: getting feedback from a wider set of users on usability and API consistency, while offering
+     enough stability to allow use in clusters running realistic loads (e.g. QA cluster for a large web service).
+   - Before graduating from alpha to beta:
+     - a feature is fully implemented; all relevant REST endpoints and kubectl commands work.  The code is tested to high standards.
+     - a feature has had an API review, though the beta period is still likely to turn up API issues not
+       found in review.
+     - the project is committed to graduating the feature to stable (typically within 3 months, but sometimes longer)
+     - has been tested to high standards and is considered safe to enable in a production cluster.
```
Comment (davidopp) [inline docs/api.md]: this is stronger than the statement a few lines earlier about "QA cluster for a large web service" -- might want to clarify

File: docs/admin/daemons.md (hunk 17 above)
Comment (davidopp) [inline docs/admin/daemons.md]: Never mind, I see you did explain this.

**Comment by bgrant0607** [comment]
As noted in #15378, we'll also need to update docs/devel/api_changes.md.

File: docs/admin/daemons.md (hunk 17 above)
Comment (lavalamp) [inline docs/admin/daemons.md]: experimental->extensions

File: docs/admin/daemons.md [hunk 20]
```diff
@@ -209,9 +209,12 @@ all or certain hosts, and when it needs to start before other pods.
...
+
+DaemonSet objects have [API version `v1alpha1`](../api.md#api-versioning).  Alpha objects may change
```
Comment (lavalamp) [inline docs/admin/daemons.md]: v1beta1, but they're turned off by default until they actually meet the beta bar.

File: docs/api.md (hunk 6 above)
Comment (lavalamp) [inline docs/api.md]: Remove the word "prefix" somehow, and add a canonical example like apis/extensions/v1beta1.

File: docs/api.md [hunk 21]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+   - *Purpose*: sharing new features with developers and eager expert users, prior to:
+     - *complete implementation*: for instance, some API operations, CLI
+     commands, or UI support may not be implemented. 
+     - *thorough API review*: the object schema may change in a later software release, and it may be necessary
+       to delete/modify/re-create objects in order to upgrade a cluster to a subsequent software release.
+       The object modification may not be easily automated, and the process may be poorly documented.
+     - *commitment to complete the feature*: the feature may be dropped entirely in a later software release.
+     - *allow rapid changes in versions*: API versions may increment faster than the minor release cadence.
+   - Alpha features are present in released binaries, but disabled by default. Enabling an Alpha feature
+     it is be necessary to make flag changes on the API server, controller, kubelet, etc..
```
Comment (lavalamp) [inline docs/api.md]: When enabling ... it may be ...

File: docs/api.md [hunk 22]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+To make it easier to extend the Kubernetes API, we are in the process of implementing [*API
+groups*](proposals/api-groups.md).  These are simply different interfaces to read and/or modify the
+same underlying resources.  The API group is specified in a REST path and in the `apiVersion` field
+of a serialized object.
+
+Currently there are two API groups in use:
+
+1. the "core" group, which is at REST path `/api/v1` and is not specified as part of the `apiVersion` field, e.g.
+   `apiVersion: v1`.
+1. the "experimental" group, which is at REST path `/apis/experimental/$VERSION`, and which uses
```
Comment (lavalamp) [inline docs/api.md]: extensions

File: docs/api.md [hunk 23]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+  `apiVersion: experimental/$VERSION` (e.g. currently `apiVersion: experimental/v1alpha1`).
+  The name `experimental` indicates that API groups themselves are currently under development.  The actual
```
Comment (lavalamp) [inline docs/api.md]: remove this now...

File: docs/api.md [hunk 24]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+  stability of the version of objects within that group is determined by its [API
+  version](#api-versioning)
+
+In the future we expect that there will be more API groups, all at REST path `/apis/$API_GROUP` and
+using `apiVersion: $API_GROUP/$VERSION`.  We expect that there will be a way for (third parties to
+create their own API groups](design/extending-api.md), and to avoid naming collisions.
```
Comment (lavalamp) [inline docs/api.md]: the way is that 3rd party group names should be formatted as a DNS name. `mygroup.mycompany.com`

File: docs/api.md [hunk 25]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+
+Not all Kubernetes clusters will support all API groups.  However, we intend that if an API server
+supports a group, it supports all the Kinds in the group, and all of them at the same version(s).
```
Comment (lavalamp) [inline docs/api.md]: No longer true-- we're turning kinds off by default until they meet the beta bar.

File: docs/api.md [hunk 26]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+
+## Enabling `experimental/v1alpha1`
+
+Enable `experimental/v1alpha1` objects by adding the following flags to your API server:
+
+  - `--runtime-config=experimental/v1alpha1=true`
+  - `--enable-experimental=true`
```
Comment (lavalamp) [inline docs/api.md]: need to double-check this.

File: docs/user-guide/jobs.md [hunk 27]
```diff
@@ -241,9 +241,11 @@ value is `Always`.)
 
 ## Caveats
 
-Job is part of the experimental API group, so it is not subject to the same compatibility
-guarantees as objects in the main API.  It may not be enabled.  Enable by setting
-`--runtime-config=experimental/v1alpha1` on the apiserver.
+Job objects are in the [`experimental` API Group](../api.md#api-groups).  This API group may or
+may not be enabled on your cluster.
+
+Job objects have [API version `v1alpha1`](../api.md#api-versioning).  Alpha objects may change
```
Comment (lavalamp) [inline docs/user-guide/jobs.md]: v1beta1

File: docs/user-guide/jobs.md [hunk 28]
```diff
@@ -241,9 +241,11 @@ value is `Always`.)
 
 ## Caveats
 
-Job is part of the experimental API group, so it is not subject to the same compatibility
-guarantees as objects in the main API.  It may not be enabled.  Enable by setting
-`--runtime-config=experimental/v1alpha1` on the apiserver.
+Job objects are in the [`experimental` API Group](../api.md#api-groups).  This API group may or
```
Comment (lavalamp) [inline docs/user-guide/jobs.md]: extensions

File: docs/admin/daemons.md (hunk 17 above)
Comment (erictune) [inline docs/admin/daemons.md]: fixed.

File: docs/admin/daemons.md (hunk 20 above)
Comment (erictune) [inline docs/admin/daemons.md]: Fixed to say `v1beta1`.

File: docs/api.md (hunk 19 above)
Comment (derekwaynecarr) [inline docs/api.md]: i prefer this sentence to the one a few earlier - the example is too vague to derive any meaning - is a web service large in number of pods, qps, data volume, etc.  seems better to just say its safe to enable in production

File: docs/api.md [hunk 29]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+   - Before graduating from alpha to beta:
+     - a feature is fully implemented; all relevant REST endpoints and kubectl commands work.  The code is tested to high standards.
+     - a feature has had an API review, though the beta period is still likely to turn up API issues not
+       found in review.
+     - the project is committed to graduating the feature to stable (typically within 3 months, but sometimes longer)
+     - has been tested to high standards and is considered safe to enable in a production cluster.
+   - The object schema may change in a later software release.  When this happens, an automatic
+     upgrade of a cluster with beta objects may not be possible.  Instead, it may be necessary
+     to delete/modify/re-create objects in order to upgrade a cluster to the new schema.  When this happens,
+     for a beta feature, we will provide documentation on how to do this.
```
Comment (derekwaynecarr) [inline docs/api.md]: when we imply automatic, do you mean no downtime, or some downtime+automated script?

File: docs/api.md [hunk 30]
```diff
@@ -53,13 +53,94 @@ In our experience, any system that is successful needs to grow and change as new
...
+   - Releases will support two consecutive versions (e.g. `v1beta1` and `v1beta2`; or `v1beta2` and `v1`)
+     for at least one minor release cycle (typically 3 months) so that users have enough time
+     to upgrade and migrate objects.
+   - TBD: do beta features enable by default?
```
Comment (derekwaynecarr) [inline docs/api.md]: i vote enable by default, but able to flag disable.

Comment (derekwaynecarr) [inline docs/api.md]: for example, in OpenShift, we will enable HPA and Job, but not Ingress for our production workloads, so it seems reasonable that others will want to make similar decisions.  since our bias is to enable more than disable, it seems fine to enable by default.

File: docs/admin/daemons.md (hunk 20 above)
Comment (erictune) [inline docs/admin/daemons.md]: Fixed to explain that they are Alpha level, but due to bug appear as v1beta1.

File: docs/api.md (hunk 6 above)
Comment (erictune) [inline docs/api.md]: did what lavalamp asked.

File: docs/api.md (hunk 16 above)
Comment (erictune) [inline docs/api.md]: added the word semantics.  I think the existing text already captures the rest of what you said.

File: docs/api.md (hunk 18 above)
Comment (erictune) [inline docs/api.md]: ok

File: docs/api.md (hunk 21 above)
Comment (erictune) [inline docs/api.md]: ok

File: docs/api.md (hunk 22 above)
Comment (erictune) [inline docs/api.md]: ok

File: docs/api.md (hunk 23 above)
Comment (erictune) [inline docs/api.md]: ok

File: docs/api.md (hunk 24 above)
Comment (erictune) [inline docs/api.md]: ack.

File: docs/api.md (hunk 25 above)
Comment (erictune) [inline docs/api.md]: removed

File: docs/api.md (hunk 26 above)
Comment (erictune) [inline docs/api.md]: @nikhiljindal should I document how to turn off individual kinds using runtime-config?

**Comment by erictune** [comment]
Okay, updated to address all comments and to match what is happening with experimental/v1alpha1 to extensions/v1beta1 change.  Humbly request we merge this and update as things evolve.

File: docs/api.md (hunk 30 above)
Comment (nikhiljindal) [inline docs/api.md]: Yes beta resources (HPA, Job and ingress) should be on by default.
Users should be able to enable/disable each extension resource. 
https://github.com/kubernetes/kubernetes/pull/15516

File: docs/api.md (hunk 26 above)
Comment (nikhiljindal) [inline docs/api.md]: We should let https://github.com/kubernetes/kubernetes/pull/15516 land. I can update the documentation in that PR if this is merged before.

File: docs/admin/daemons.md (hunk 20 above)
Comment (lavalamp) [inline docs/admin/daemons.md]: missing push?

File: docs/admin/daemons.md (hunk 17 above)
Comment (bgrant0607) [inline docs/admin/daemons.md]: Need to push?

**Comment by bgrant0607** [comment]
Waiting for recent changes to be pushed.

**Comment by erictune** [comment]
Pushed.

**Comment by erictune** [comment]
Okay, I reworked this change.  I put detailed developer-centric descriptions of the levels in docs/devel/api_changes.md, as bgrant0607 requested.  And I referenced that from docs/api.md.   That allowed me to make the description in docs/api.md much shorter.  Also, after talking to @quinton-hoole , I put on my SRE hat and tried to give suggestions on in what types of clusters you would want to use the different feature levels.  

I will squash when there are no more comments.

File: docs/api.md (hunk 19 above)
Comment (erictune) [inline docs/api.md]: I reworked the text and dropped the offensive sentence.

File: docs/api.md (hunk 29 above)
Comment (erictune) [inline docs/api.md]: I reworked the text, and now specify downtime as the expected case.

File: docs/api.md (hunk 30 above)
Comment (erictune) [inline docs/api.md]: Yes, beta enabled by default.

**Comment by erictune** [comment]
@karlkfi we use the "e2e-not-required" label on docs-only changes to spare the e2e system excessive load.  if you want, you can teach @k8s-teamcity-mesosphere to ignore PRs with this label.

**Comment by erictune** [comment]
rewrote again.

File: docs/devel/api_changes.md [hunk 31]
```diff
@@ -535,6 +535,69 @@ The API spec changes should be in a commit separate from your other changes.
...
+    need not have had an *API review* (an intensive and targeted review of the API, on top of a normal
+    code review)
+  - Upgradeability: the object schema and semantics may change in a later software release, without
+    any provision for preserving objects in an existing cluster;
+    removing the upgradability concern allows developers to make rapid progress; in particular,
+    API versions can increment faster than the minor release cadence and the developer need not
+    maintain multiple versions; developers should still increment the API version when object schema
+    or semantics change in an [incompatible way](#on-compatibility)
+  - Cluster Reliability: because the feature is relatively new, and may lack complete end-to-end
+    tests, enabling the feature via a flag might expose bugs with destabilize the cluster (e.g. a
```
Comment (lavalamp) [inline docs/devel/api_changes.md]: s/with/which/

File: docs/devel/api_changes.md [hunk 32]
```diff
@@ -535,6 +535,69 @@ The API spec changes should be in a commit separate from your other changes.
...
+    and lack of long-term support and lack of upgradability.
+- Beta level:
+  - Object Versioning: API version name contains `beta` (e.g. `v2beta3`)
+  - Availability: in official Kubernetes releases, and enabled by default
+  - Audience: users interested in providing feedback on features
+  - Completeness: all API operations, CLI commands, and UI support should be implemented; end-to-end
+    tests complete; the API has had a thorough API review and is thought to be complete, though use
+    during beta may frequently turn up API issues not thought of during review
+  - Upgradeability: the object schema and semantics may change in a later software release; when
+    this happens, an upgrade path will be documentedr; in some cases, objects will be automatically
```
Comment (lavalamp) [inline docs/devel/api_changes.md]: s/documentedr/documented/

**Comment by lavalamp** [comment]
LGTM, I think the typos are minor & can fix in followup

**Comment by bgrant0607** [comment]
I agree we should merge and iterate.

After this merges, I can update the text I wrote here:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api_changes.md#incompatible-api-changes

**Comment by karlkfi** [comment]
It's less of a bot and more of a TeamCity plugin someone else wrote, but yes, we could make it smarter. 

-Karl

**Comment by bgrant0607** [comment]
Please cherrypick this into release-1.1.

**Comment by erictune** [comment]
all content from this PR is in release-1.1
